### PR TITLE
Add rmsnorm kernel

### PR DESCRIPTION
**Pull request by rahulbatra85**
Adds forward kernel for RMSNorm

### CODE REVIEW

File: python/perf-kernels/rmsnorm.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+        triton.Config({}, num_warps=16, num_stages=1),
+        triton.Config({}, num_warps=4, num_stages=1),
+        triton.Config({}, num_warps=8, num_stages=1),
+        triton.Config({}, num_warps=16, num_stages=1),
+        triton.Config({}, num_warps=4, num_stages=1),
+        triton.Config({}, num_warps=8, num_stages=1),
+        triton.Config({}, num_warps=16, num_stages=1),
+    ]
+
+def get_hip_autotune_config():
```
Comment (jtang10) [inline python/perf-kernels/rmsnorm.py:29]: Nothing wrong here. Just wondering if these configs are comprehensive enough to cover the typical use cases you have encountered in actual models

Comment (rahulbatra85) [reply python/perf-kernels/rmsnorm.py:29]: Good question, I just came up with these on what I thought made sense. Other than that, no systematic way to come up with this. I was thinking may be I can add an argument to take in a custom config file. That way one doesn't need to touch the code in this file if some other configs need to be benchmarked.

Actually, I noticed a small bug for CUDA. There are repeated entries.

**Comment by brunomazzottiamd** [comment]
There's a [README](https://github.com/ROCm/triton/blob/main_perf/python/perf-kernels/README.md) in the `perf-kernels` directory. It'd nice to add a line about **RMSNorm** to the README.

**Comment by brunomazzottiamd** [comment]
**"pre-commit (code formatting)"** check isn't passing. It's related to Triton coding standards. A couple of weeks ago, I accidentally merged a PR that wasn't passing this check. After that, Michael made passing this check mandatory for merging. I think it's a good thing to do if someday we may want to upstream our kernels.

You can `pip install pre-commit` and then `pre-commit install` in the root folder of your Triton repository. Then your diff hunks will be checked against https://github.com/triton-lang/triton/blob/main/.pre-commit-config.yaml rules. You can find more info at https://pre-commit.com/, it's a very interesting tool.

File: python/perf-kernels/rmsnorm.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+    for row_idx in tl.range(row_start, n_rows, row_step):
+        row_start_ptr = input_ptr + row_idx * input_row_stride
+        col_offsets = tl.arange(0, BLOCK_SIZE)
+        mask = col_offsets < n_cols
+        
+        row = tl.load(row_start_ptr + col_offsets, mask=mask, other=0.0)
+        g = tl.load(g_ptr + col_offsets, mask=mask, other=0.0)
+        row_norm = row * row #square each value
+        row_norm = tl.sum(row_norm, axis=-1) #sum across columns(axis=-1)
+        row_norm = row_norm / n_cols #divide by n_cols
```
Comment (brunomazzottiamd) [inline python/perf-kernels/rmsnorm.py:66]: This is just a curiosity of mine, it isn't a blocking comment at all.

Do you know how `row_norm = row_norm / n_cols` is being lowered by the compiler to assembly? As far as I can see `n_cols` is a scalar. I'm curious if we are performing `BLOCK_SIZE` divisions or `BLOCK_SIZE` multiplications and what approach is the faster.

In my mind `reciprocal_n_cols = 1 / n_cols` and `row_norm = row_norm * reciprocal_n_cols` is one scalar division followed by a `BLOCK_SIZE` vector multiplication. Probably this won't make any difference because the kernel is memory bound.

File: python/perf-kernels/rmsnorm.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+    row_step = tl.num_programs(0)
+    for row_idx in tl.range(row_start, n_rows, row_step):
+        row_start_ptr = input_ptr + row_idx * input_row_stride
+        col_offsets = tl.arange(0, BLOCK_SIZE)
+        mask = col_offsets < n_cols
+        
+        row = tl.load(row_start_ptr + col_offsets, mask=mask, other=0.0)
+        g = tl.load(g_ptr + col_offsets, mask=mask, other=0.0)
+        row_norm = row * row #square each value
+        row_norm = tl.sum(row_norm, axis=-1) #sum across columns(axis=-1)
```
Comment (brunomazzottiamd) [inline python/perf-kernels/rmsnorm.py:65]: For **RMSNorm**, is it always safe to accumulate FP16 values with a FP16 accumulator? Do we need a FP32 accumulator here? I'm supposing that `tl.sum` of FP16's is also a FP16.

For instance, if we square 7 three-digit numbers and then add them together we're already overflowing FP16.

I'm not that familiar with normalization layers, maybe I'm just too conservative and I'm thinking too much in the general case.

File: python/perf-kernels/rmsnorm.py [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+def test_rmsnorm(M, N):
+    torch.manual_seed(0)
+    x = torch.randn(M, N, device='cuda')
+    #print(x)
+    y_triton = rmsnorm(x)
+    #print(y_triton)
+
+    rms_norm = torch.nn.RMSNorm(N, device='cuda')
+    y_torch = rms_norm(x)
+    #print(y_torch)
```
Comment (brunomazzottiamd) [inline python/perf-kernels/rmsnorm.py:125]: Can we remove these debug print statements before merging?

File: python/perf-kernels/rmsnorm.py (hunk 2 above)
Comment (rahulbatra85) [reply python/perf-kernels/rmsnorm.py:66]: @brunomazzottiamd  I have the TTGIR and LLIR captured here. Assembly was a lot more lines.
https://github.com/orgs/ROCm/projects/28/views/2?pane=issue&itemId=67348666

By the way, your question made me think and I realized that I can combine the tl.sqrt and the divide on the next line and use rsqrt. I guess I still will need to do an op instead of divide do a multiply.

Comment (brunomazzottiamd) [reply python/perf-kernels/rmsnorm.py:66]: That's great! Can you check if `rsqrt` improve performance? Inverse square root reminded of https://en.wikipedia.org/wiki/Fast_inverse_square_root...

**Comment by brunomazzottiamd** [comment]
Seeing this PR reminded that I added a kernel without a benchmark and without a test! Shame on me! I think adding benchmark and correctness test must be mandatory from now on. Reviewers should reject PRs lacking these features.

**Comment by rahulbatra85** [comment]
> **"pre-commit (code formatting)"** check isn't passing. It's related to Triton coding standards. A couple of weeks ago, I accidentally merged a PR that wasn't passing this check. After that, Michael made passing this check mandatory for merging. I think it's a good thing to do if someday we may want to upstream our kernels.
> 
> You can `pip install pre-commit` and then `pre-commit install` in the root folder of your Triton repository. Then your diff hunks will be checked against https://github.com/triton-lang/triton/blob/main/.pre-commit-config.yaml rules. You can find more info at https://pre-commit.com/, it's a very interesting tool.

@brunomazzottiamd  yeah, I have used this pre-commit git hooks tool before with other projects. Didn't realize we are using it here as well. Thanks for the tip!

File: python/perf-kernels/rmsnorm.py [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+    key=['n_rows', 'n_cols'],
+    use_cuda_graph=True
+)
+@triton.jit
+def rms_kernel(output_ptr, input_ptr, g_ptr, input_row_stride, output_row_stride, n_rows, n_cols, epsilon, BLOCK_SIZE: tl.constexpr):
+    row_start = tl.program_id(0)
+    row_step = tl.num_programs(0)
+    for row_idx in tl.range(row_start, n_rows, row_step):
+        row_start_ptr = input_ptr + row_idx * input_row_stride
+        col_offsets = tl.arange(0, BLOCK_SIZE)
```
Comment (brunomazzottiamd) [inline python/perf-kernels/rmsnorm.py:59]: `col_offsets` and `mask` are loop invariants. Does it make any difference moving them outside the loop?

**Comment by rahulbatra85** [comment]
> There's a [README](https://github.com/ROCm/triton/blob/main_perf/python/perf-kernels/README.md) in the `perf-kernels` directory. It'd nice to add a line about **RMSNorm** to the README.

Done!

File: python/perf-kernels/rmsnorm.py (hunk 4 above)
Comment (rahulbatra85) [reply python/perf-kernels/rmsnorm.py:125]: Fixed

File: python/perf-kernels/rmsnorm.py (hunk 3 above)
Comment (rahulbatra85) [reply python/perf-kernels/rmsnorm.py:65]: I need to think about this case a bit more.

File: python/perf-kernels/rmsnorm.py (hunk 2 above)
Comment (rahulbatra85) [reply python/perf-kernels/rmsnorm.py:66]: Added rsqrt

File: python/perf-kernels/rmsnorm.py (hunk 3 above)
Comment (brunomazzottiamd) [reply python/perf-kernels/rmsnorm.py]: Let me try to give you some food for thought...

### Use a wider data type as accumulator

Our GEMM kernel does this:
```python
# INT32 accumulator for INT8 data and FP32 accumulator for everything else.
acc_dtype = tl.float32 if a_ptr.type.element_ty != tl.int8 else tl.int32
accumulator = tl.zeros((BLOCK_SIZE_M, BLOCK_SIZE_N), dtype=acc_dtype)
```

You can do something similar. I think it's safe to accumulate using FP32.

### Use a scale factor

You can scale down the numbers before squaring them and scale up the mean of squares afterwards. Here is a NumPy prototype of the idea:
```python
import numpy as np


def mean_sqr(x):
    return np.sum(x * x) / len(x)


# Scale factor can be a constant or computed on the fly.
def mean_sqr_with_scale_factor(x, scale_factor):
    x = (1 / scale_factor) * x
    mean_sqr = np.sum(x * x) / len(x)
    return (scale_factor * scale_factor) * mean_sqr


def compute_scale_factor(x):
    max_x = np.max(np.abs(x))
    return np.exp(np.floor(np.log(max_x))) if max_x > 0 else 1


np.random.seed(42)
x = np.random.uniform(size=4096, low=-500.0, high=500.0).astype(np.float32)
print(mean_sqr(x))
print(mean_sqr_with_scale_factor(x, compute_scale_factor(x)))
```

Larger scale factors reduce the risk of overflow but increase the chance of losing precision. Smaller scale factors retain precision but may not prevent overflow effectively if the numbers are too large.

### Inspect PyTorch implementation

Try to find out what PyTorch is doing. If PyTorch doesn't care about sum of squares overflow then we can follow it and doesn't care as well.

**Review by brunomazzottiamd — APPROVED** [review]

**Review by micmelesse — CHANGES_REQUESTED** [review]
Can we add tests for this kernel?

**Comment by micmelesse** [comment]
@rahulbatra85 There are failures in the RMS prop code

**Comment by rahulbatra85** [comment]
> @rahulbatra85 There are failures in the RMS prop code

yeah, looking.

**Comment by rahulbatra85** [comment]
> > @rahulbatra85 There are failures in the RMS prop code
> 
> yeah, looking.

@micmelesse  ah ok, so rms norma layer was added in Pytorch  starting with version 2.4. Which docker image does the CI use?

**Comment by micmelesse** [comment]
@rahulbatra85 It's `rocm/pytorch:rocm6.0.2_ubuntu22.04_py3.10_pytorch_2.1.2` you can probably update it to the newest one

**Comment by rahulbatra85** [comment]
> @rahulbatra85 It's `rocm/pytorch:rocm6.0.2_ubuntu22.04_py3.10_pytorch_2.1.2` you can probably update it to the newest one

@micmelesse  It's passing with new docker image.

**Review by micmelesse — APPROVED** [review]
